File: back/src/main/java/com/openclassrooms/mddapi/service/TopicService.java
package com.openclassrooms.mddapi.service;

import com.openclassrooms.mddapi.DTO.TopicResponse;
import com.openclassrooms.mddapi.model.Topic;

import java.util.List;

public interface TopicService {
    List<Topic> getAllTopics();

    void subscribe(String username, Long topicId);

    void unsubscribe(String username, Long topicId);

    List<TopicResponse> getAllTopicsWithSubscriptionStatus(String username);

}

File: front/src/app/services/topic.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Topic, TopicResponse } from '../models/topic.model';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root',
})
export class TopicService {
  
  private apiUrl =  `${environment.apiUrl}/topics`;

  constructor(private http: HttpClient) {}

  getAllTopics(): Observable<Topic[]> {
    return this.http.get<Topic[]>(this.apiUrl);
  }

  subscribeToTopic( topicId: number): Observable<string> {
    return this.http.post(`${this.apiUrl}/subscribe/${topicId}`, null, {
      responseType: 'text',
    });
  }
  
  unsubscribeFromTopic( topicId: number): Observable<string> {
    return this.http.post(`${this.apiUrl}/unsubscribe/${topicId}`, null, {
      responseType: 'text',
    });
  }

  getAllTopicsWithStatus(): Observable<TopicResponse[]>{
    return this.http.get<TopicResponse[]>(`${this.apiUrl}/subscriptions`);
  }
  
  
}
